pytransform3d/transform_manager/_temporal_transform_manager.py

```diff
@@ -96,37 +105,61 @@ def __init__(self, time, pqs):
             raise ValueError("`pqs` matrix shall have 7 columns.")
 
     def as_matrix(self, query_time):
+        """Get transformation matrix at given time.
+
+        Parameters
+        ----------
+        query_time : Union[float,array-like shape (...)]
+            Query time
+
+        Returns
+        -------
+        A2B_t : array, shape (4, 4) or (..., 4, 4)
+            Homogeneous transformation matrix at given time. . or times
+        """
         pq = self._interpolate_pq_using_sclerp(query_time)
-        return transform_from_pq(pq)
+        transforms = transforms_from_pqs(pq)
+        transforms = np.squeeze(transforms)  # to keep the external API
+        return transforms
 
     def check_transforms(self):
         self._pqs[:, 3:] = norm_vectors(self._pqs[:, 3:])
         return self
 
     def _interpolate_pq_using_sclerp(self, query_time):
+        query_time_arr = np.atleast_1d(query_time)
+
         # identify the index of the preceding sample
-        idx_timestep_earlier_wrt_query_time = np.argmax(
-            self.time >= query_time) - 1
+        idxs_timestep_earlier_wrt_query_time = np.searchsorted(
+            self.time, query_time_arr, side='right'
+        ) - 1
 
-        # deal with first timestamp
-        idx_timestep_earlier_wrt_query_time = max(
-            idx_timestep_earlier_wrt_query_time, 0)
+
+        # deal with first and last timestamp
+        min_index = 0
+        max_index = self.time.shape[0] - 2
+        idxs_timestep_earlier_wrt_query_time = np.clip(
+            idxs_timestep_earlier_wrt_query_time,
+            min_index,
+            max_index
+        )
 
         # dual quaternion from preceding sample
-        t_prev = self.time[idx_timestep_earlier_wrt_query_time]
-        pq_prev = self._pqs[idx_timestep_earlier_wrt_query_time, :]
-        dq_prev = dual_quaternion_from_pq(pq_prev)
+        t_prev = self.time[idxs_timestep_earlier_wrt_query_time]
+        pq_prev = self._pqs[idxs_timestep_earlier_wrt_query_time, :]
+        dq_prev = dual_quaternions_from_pqs(pq_prev)
 
         # dual quaternion from successive sample
-        t_next = self.time[idx_timestep_earlier_wrt_query_time + 1]
-        pq_next = self._pqs[idx_timestep_earlier_wrt_query_time + 1, :]
-        dq_next = dual_quaternion_from_pq(pq_next)
+        t_next = self.time[idxs_timestep_earlier_wrt_query_time + 1]
+        pq_next = self._pqs[idxs_timestep_earlier_wrt_query_time + 1, :]
+        dq_next = dual_quaternions_from_pqs(pq_next)
 
         # since sclerp works with relative (0-1) positions
         rel_delta_t = (query_time - t_prev) / (t_next - t_prev)
-        dq_interpolated = dual_quaternion_sclerp(dq_prev, dq_next, rel_delta_t)
-
-        return pq_from_dual_quaternion(dq_interpolated)
+        dqs_interpolated = dual_quaternions_sclerp(
+            dq_prev, dq_next, rel_delta_t)
+        res = pqs_from_dual_quaternions(dqs_interpolated)
+        return res
```

Comment:
If we pass a float query_time, this should return an array of shape (4, 4).

My solution to this would be to put as a first line of the function `query_shape = np.asarray(query_time).shape` and reshape with `res.reshape(*(list(query_shape) + [4, 4]))`.